### PR TITLE
[Form] provide a useful message when extension types don't match

### LINKED ISSUE #36610: [Form] DependencyInjectionExtension Type extension exception
**Symfony version(s) affected**: 4.2.0 and above (currently up until 5.1)

**Description**  
The `getTypeExtensions($name)` method of \Symfony\Component\Form\Extension\DependencyInjection\DependencyInjectionExtension throws a strange exception message:

> The extended type specified for the service "0" does not match the actual extended type. Expected "Vendor\Bundle\AcmeBundle\Form\FooType", given "Vendor\Bundle\AcmeBundle\Form\BarType".

See the hightlighted code lines of the `getTypeExtensions()` method in the additional context below, the `$serviceId` variable contains the integer key of the items of the `$this->typeExtensionServices[$name]` iterable. Apparently, this has been set to an actual service ID by a service ID guesser before Symfony went full FQCN and now it just remains a numeric key. But the thrown exception message has not been adjusted to the new situation. This results is a very unclear error.

**How to reproduce**  
Have a Type extension's `getExtendedType`(`s`) return invalid type(s)

**Possible Solution**  
Clarify the exception message with useful data. Also the key in the loop is therefore probably no longer needed.

**Additional context**  
https://github.com/symfony/symfony/blob/4170346ea736d90de32384cf4fea45f2aa0f4d73/src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php#L64-L78

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #36610
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php [hunk 1]
```diff
@@ -68,7 +68,7 @@ public function getTypeExtensions($name)
 
                 // validate the result of getExtendedTypes()/getExtendedType() to ensure it is consistent with the service definition
                 if (!\in_array($name, $extendedTypes, true)) {
-                    throw new InvalidArgumentException(sprintf('The extended type specified for the service "%s" does not match the actual extended type. Expected "%s", given "%s".', $serviceId, $name, implode(', ', $extendedTypes)));
+                    throw new InvalidArgumentException(sprintf('The extended type "%s" specified for the type extension class "%s" does not match any of the actual extended types (["%s"]).', $name, \get_class($extension), implode('", "', $extendedTypes)));
```
Comment (javiereguiluz) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php:71]: This error message is a bit confusing to me. Could we use the common Symfony pattern for errors like this? (Maybe it doesn't make sense, I don't know!) E.g.

```
The extended type of class "%s" must be one of ["%s"], "%s" given.
```

**Comment by 7ochem** [comment]
Resolves #36610

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 1 above)
Comment (xabbuh) [reply src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php:71]: The suggested wording conveys the assumption that the class returns unexpected data while the extension class currently assumes that the service configuration is incorrect. I don't know what would be better. 🤷‍♂️

Comment (javiereguiluz) [reply src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php:71]: OK, let's keep then message then. Thanks!

**Review by yceruto — APPROVED** [review]

**Review by nicolas-grekas — APPROVED** [review]

**Review by OskarStark — APPROVED** [review]

**Review by 7ochem — APPROVED** [review]

**Pull request merged**